### ISSUE TITLE
Simplifying integration with hook registration in Dear ImGui?

### ISSUE THREAD

**Issue by ocornut** [sammyfreg/netImgui#9]
I added NetImGui to a test project today.
For reference, it took me 45 minutes to figure things out from scratch (quite good!), and the following lines of codes.

```
#pragma comment(lib, "Ws2_32")
#include <NetImgui_Api.h>
```
```
bool ret = NetImgui::Startup();
IM_ASSERT(ret);
if (NetImgui::ConnectToApp("Sandbox Client", "localhost", 8888U, true))
{
    ImFontAtlas* atlas = ImGui::GetIO().Fonts;
    app->NewFrame(app); // FIXME: enforce texture upload. Should do the upload lazily in the main loop instead
    unsigned char* pixels = NULL;
    atlas->GetTexDataAsRGBA32(&pixels, NULL, NULL, NULL);
    NetImgui::SendDataTexture(
             (uint64_t)atlas->TexID, atlas->TexPixelsRGBA32, 
             atlas->TexWidth, atlas->TexHeight, NetImgui::eTexFormat::kTexFmtRGBA8);
}
```
```
if (NetImgui::IsConnected())
    NetImgui::NewFrame();
else
    ImGui::NewFrame();
```
```
if (NetImgui::IsConnected())
    NetImgui::EndFrame();
else
    ImGui::Render();
```

As discussed in #8 i believe that NetImgui could handle the texture aspect automatically.

**My suggestion:**

I could add custom hook registrations in ImGui that would allow NetImgui to register function to be called in NewFrame() / EndFrame() / Render(). This way, the user wouldn't even have to call `NetImgui::NewFrame();` or `NetImgui::EndFrame();`.

Do you think that would that be useful and you can use that?

The TestEngine already has custom hooks that ImGui calls and I would simply replace those specialized hooks with a more generic system, so it's not much work on ImGui side

**Comment by sammyfreg** [comment]
```
if (NetImgui::IsConnected())
    NetImgui::NewFrame();
else
    ImGui::NewFrame();
```

This is not needed anymore, you can keep a single call to `NetImgui::NewFrame()/NetImgui:EndFrame()` and things will be handled properly.

The custom hook idea is definitly interesting. We would have to think about how things are handled underneath. You'll notice that inside my `NetImGui::NewFrame` I call `Imgui::NewFrame` myself, on either the original context, or a cloned context. Moreover, if user handles it, I can skip frame rendering, when we haven't received a request for a new frame from the server yet.

Speaking of cloned context, this is one of my main **Dear ImGui** version "*compatibility issue*". It's no a big problem, but when I update **Dear ImGui** version, I make sure that the cloning handles the new and removed data member. It could be nice if a function directly in **Dear ImGui** could handle it.

**Comment by sammyfreg** [comment]
> app->NewFrame(app); // FIXME: enforce texture upload. Should do the upload lazily in the main loop instead

This can actually be skipped, there's no need to assign it before connection or first frame. You can upload the font texture where you create it normally without issue. It should just use a blank texture until it is assigned.

I will still look into auto-upload as suggested in #8

**Comment by ocornut** [comment]
> This is not needed anymore, you can keep a single call to NetImgui::NewFrame()/NetImgui:EndFrame() and things will be handled properly.

Well if we are aiming to make magic happen it would be even better to not have to make this change at all :)

I added the hooks:
https://github.com/ocornut/imgui/commit/a38c6dfcc83fba7166c953e54cfbf8dceba7be0b
So using imgui_internal.h if you just add yourself to `g.Hooks` you should be able to get this to work automatically.

I don't really understand why/how you are doing this clone thing but I see that the `ContextClone()` function effectively only copies the contents of ImGuiIO + ImGuiStyle. I would just memcpy the ImGuiIO structure and then override the few members that needs it.

**Comment by sammyfreg** [comment]
It looks like it'll work nicely, thanks! I am a bit busy this month, so I will look into this in the comings weeks.

As for the cloning, It allows user to leave the original context intact while also importing its config settings. Useful when user want to continue displaying content locally and remotely for example, so 2 distincts context. I initially thought about a simple memcpy, but there's a non pod object and some internal state data in `ImGuiIO` that I didn't want to touch. I suppose I could still do a memcpy, but stop before the internal data.

**Comment by ocornut** [comment]
I think you can treat ImGuiIO as mostly a POD, notice that it's constructor starts with `memset(NULL, ...)`.
Using `=` operator to copy one should work. There's once 1 field which is ImVector<> and frankly it should be empty when you clone anyway, but using `=` operator will copy it if not.

**Comment by sammyfreg** [comment]
I am not sure where you'd prefer to discuss support for the hook, here or on a new issue over at Dear ImGui ?

I took a quick look at the hook (I am not very far into it), and had 2 issues to address.

**1. Must include `imgui_internal.h`**
Will the final version still needs this include, or will it be moved  in the public facing `imgui.h` ? Right now, I require the **NetImgui** users to only provide me with the imgui.h include (inside `netimgui_config.h`). I would rather avoid asking the user to also give me the path to `imgui_internal.h`

**2. Context in use and hook callback**
My hook callback can potentially change the current context before we start drawing. It might need to replace it with either the 'cloned' context or an 'empty' context (used to receive and discard drawing when server is not expecting another frame just yet).

Easy fix :
```
void ImGui::NewFrame()
{
    IM_ASSERT(GImGui != NULL && "No current context. Did you call ImGui::CreateContext() and ImGui::SetCurrentContext() ?");
    CallContextHooks(&GImGui, ImGuiContextHookType_NewFramePre); *** (  Call hook ) *** 
    ImGuiContext& g = *GImGui;  *** ( Moved after the hook callback)  ***
    // ...
    g.Time += g.IO.DeltaTime; 
    g.WithinFrameScope = true;

```

**Comment by sammyfreg** [comment]
Hook registration support added in dev branch, with 974e5f9807fbd3e2e9fed3c77a2576e98e5c1ba4. Will be available in main branch on next release.
